### PR TITLE
Support Nemotron 3.5 Lightning Q8 GGUF conversion

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

- add a reusable GGUF architecture-adapter registry so generic import code owns dispatch and completeness while `nemotron_h_moe` owns its schedule, tensor names, transforms, qtype policy, tokenizer contract, and MTP semantics
- convert the pinned Nemotron 3.5 Lightning Q8_0 artifact to an executable ONNX package while preserving Q8 blocks as `MatMulNBits(bits=8, block_size=32)`, including leading-axis routed-expert expansion without dequantization
- reconstruct the exact GPT-2/Pixtral tokenizer engine, keep tokenizer and model-runtime special-token contracts explicit, and add a reproducible fresh-process direct-ORT acceptance runner

## Stack and lineage

This PR is linearly restacked on exact #487 head `8ccfc25977d9551c7dfc07088eea1827a5fcb8b8` (itself based on current main `cf77d0f7fb28adebbeb3aaecb25fe2e51ee0ddd0`). Exact PR head is `b0b94f92cd39aea883bdbc008a3ff3ff31b56121`; `8ccfc259...HEAD` contains exactly one commit, `Support Nemotron 3.5 Q8 GGUF conversion`.

The parent inherits merged #500 skills once and retains #498 once. This PR has no `.agents/` or `.github/` delta and does not duplicate either dependency. The only rebase adaptation updates a test import to the current `mobius.integrations.transformers` location introduced by the ecosystem refactor.

## Exact architecture and mapping contract

The adapter validates the complete pinned tensor table before graph construction:

- 417 source tensors
- 401 mapped backbone sources and 16 explicit block-52 MTP exclusions
- 52 backbone layers: 23 Mamba + 23 MoE + 6 attention
- 6,243 logical targets, including 6,006 exact Q8 targets
- 18,255 mapped weight initializers; missing, unexpected, duplicate, or unmapped targets fail the build
- `ssm_a -> log(-A)`, A/D/grouped-norm squeeze/flatten, conv-axis expansion, inverse llama.cpp Q/K row permutation, and direct expert-axis slicing are shape-validated

All Nemotron constants and error prose live in `src/mobius/integrations/gguf/_nemotron_h_moe.py`; model/component factory changes are source-format agnostic and the existing float/Safetensors graph remains unchanged. Every registered adapter must implement source validation, and generic validation dispatch is explicitly typed.

## Pinned real-artifact acceptance

Source: `unsloth/NVIDIA-Nemotron-3.5-Lightning-30B-A3B-GGUF@f2d3fe3694501008786e81e5f20360cbf715496a`

- file: `NVIDIA-Nemotron-3.5-Lightning-30B-A3B-Q8_0.gguf`
- size: 35,004,643,392 bytes
- SHA-256: `dc5276dd0619c04e277504d2358a793e31ccbe39e894d767d0d14f2a221e2ca4`
- inventory: 32,904,044,544 Q8_0 parameters, 8,877,632 F32 parameters, 344,064 BF16 MTP parameters

The full build/save/reload/prefill/cached-decode acceptance was rerun from exact head `b0b94f92cd39` with this worktree first on `PYTHONPATH` and an isolated approved-feed ONNX Runtime 1.28.0 install:

- build: 210.616 s; save: 127.958 s
- package: 36,920,438,736 bytes
- peak build working set: 52,957,401,088 bytes
- 6,005 `MatMulNBits` + 1 `GatherBlockQuantized`
- 6,006 Q8 weight initializers and 18,255 mapped weight initializers
- no `QuantizeLinear` or `DequantizeLinear`

A fresh process loaded the saved external-data package in ONNX Runtime 1.28.0 (`45de2a8b06`) on `CPUExecutionProvider`, ran nonzero unpadded prefill, independently ran right-padded prefill with an explicit mask, and executed every cached decode step:

- session load: 25.607 s
- five-token prefill: 97.514 s
- seven cached decode calls: 52.553, 52.440, 52.230, 53.129, 52.512, 52.118, 52.823 s
- cached decode throughput: 0.01903 steps/s
- peak runtime working set: 38,146,613,248 bytes
- right-padded real-token logits max absolute difference: `0.0` (`atol=1e-5`)

Independent llama.cpp reference: commit `9d57ce456c94d241dde672b2db9cf18879766568`, MSVC 19.44.35228.0, CPU only:

```text
llama-server -m NVIDIA-Nemotron-3.5-Lightning-30B-A3B-Q8_0.gguf -c 128 -t 12 -tb 12 -b 64 -ub 64 -ngl 0 --host 127.0.0.1 --port 18081 --no-warmup
POST /completion
{"prompt":"The capital of France is","n_predict":8,"temperature":0,"seed":1,"cache_prompt":false,"n_probs":1}
```

- prompt IDs: `[1784, 8961, 1307, 5498, 1395]`
- generated IDs: `[6993, 1046, 1256, 1010, 1784, 8961, 1307, 10787]`
- generated text: ` Paris.  \nThe capital of Germany`
- llama.cpp throughput: 5.58 prompt tok/s, 5.95 generated tok/s
- separate llama.cpp CLI peak working set: 16.81 GiB

The exact-head direct-ORT run matched every generated ID and the decoded text exactly.

## Tokenizer and special-token contract

The GGUF tokenizer is selected only for exact `model=gpt2, pre=pixtral` metadata and matches the pinned official tokenizer at `nvidia/NVIDIA-Nemotron-3.5-Lightning-30B-A3B-BF16@d468880b6ad3c6e0d21377ce7242adaea4cc884d` across vocabulary, added-token flags, pre-tokenizer, decoder, post-processor, encode IDs, and decode text.

- 131,072-token hash: `4999709474e3c967358c1f1199b6be65fb9055d3eb59e0cd387f9e7077fc40ed`
- 269,443-merge hash: `b1b0165185b1925118c2f7b1e978439b02010c3a420ebfec5c19a093a0d9b4cb`
- official tokenizer asset SHA-256: `623c34567aebb18582765289fbe23d901c62704d6518d71866e0e58db892b5b7`
- GGUF BOS/EOS/PAD metadata: `1/11/999`
- model-runtime BOS/EOS/PAD: `1/2/0`
- generation stop IDs: `[2, 11]`
- tokenizer asset EOS/PAD role: `<|im_end|>` ID 11

PAD 999 is rejected because it names `<SPECIAL_999>`. The GGUF-embedded chat template differs from the pinned official asset, so the recipe verifies immutable official sidecar hashes rather than silently treating it as authoritative. Refreshed #487 sentinel filtering and EOS/EOT/EOM union semantics remain intact for the generic path.

## Explicit limits

- Q4_K_M remains rejected: its large sources are Q5_0/Q5_1, not a preservable Q4-only payload.
- MXFP4_MOE remains rejected: it contains 12.772B Q5_1 parameters without a validated preserved runtime mapping.
- split BF16 GGUF remains rejected by the generic shard-assembly guard.
- dequantize/requantize is not described as preservation.

## Validation on exact head

- focused GGUF/Nemotron/build-graph suite: `1,692 passed, 47 skipped`
- pinned real-header + official-tokenizer parity: `2 passed`
- canonical non-integration suite: `4,130 passed, 64 skipped`, plus four failures in untouched upstream-asset patch fixtures
- those same four selectors reproduce on detached exact base `8ccfc259`; excluding only them gives `4,130 passed, 64 skipped`
- `lintrunner -a`: clean
- all 15 changed Python files pass Ruff format check and lint; `git diff --check` is clean
- the Windows all-files formatter adapter still hits WinError 206 from expanded command length; it produced no PR change
- independent specialist review of `8ccfc259...b0b94f92`: no significant findings
- all three existing automated review threads remain replied to and resolved

## Waivers

- **ORT GenAI:** generic metadata cannot bind arbitrary non-KV convolution/SSM recurrent-state inputs. This is a graph-structural runtime limitation, so direct ORT cached generation is the acceptance path.
- **CUDA Q8 runtime:** the 36.9 GB package cannot fit the available 8 GB GPU. CPU acceptance uses the available hardware and records full semantic/memory evidence.
- **Foundry Local / Q8 Olive:** not run for this 36.9 GB GGUF follow-up. Exact parent #487 `8ccfc259` validates the official model's CPU/CUDA cached generation and fresh Olive Q4 package; this PR specifically validates the preserved source-Q8 ONNX package through direct ORT.

### CODE REVIEW

File: src/mobius/integrations/gguf/_architecture.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    return adapter_type
+
+
+def _load_builtin_adapters() -> None:
+    global _BUILTINS_LOADED
+    if _BUILTINS_LOADED:
+        return
+    for module_name in _BUILTIN_MODULES:
+        import_module(module_name)
+    _BUILTINS_LOADED = True
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_architecture.py:152]: `_BUILTINS_LOADED` is used as the idempotence guard in `_load_builtin_adapters()`: the function reads it before importing built-ins and sets it to `True` after registration. Keeping the guard prevents duplicate adapter registration on subsequent lookups, so no change is needed.

File: src/mobius/integrations/gguf/_architecture.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    architecture: str
+    model_type: str
+
+    def __init__(self, model: GGUFModel) -> None:
+        self.model = model
+
+    def validate_model(self, *, source: str) -> None:
+        """Validate the source tensor table and supported quantization."""
+
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_architecture.py:91]: Fixed in `4e9ab2e`: the base `validate_model` now raises `NotImplementedError`, so every registered adapter must provide source validation. Added a regression test for that contract.

File: src/mobius/integrations/gguf/_builder.py [hunk 3]
```diff
@@ -176,21 +106,19 @@ def _preflight_hf_gguf(api: HfApi, repo_id: str, filename: str) -> None:
...
-        )
+        logger.debug("Hub GGUF architecture preflight for %s: %s", source, architecture)
 
 
-def _validate_gguf_model(gguf_model, *, source: str) -> None:
+def _validate_gguf_model(gguf_model, *, source: str):
     """Validate a parsed GGUF before config extraction or graph construction."""
+    from mobius.integrations.gguf._architecture import create_architecture_adapter
+
     split_count = int(gguf_model.get_metadata("split.count", 1))
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/integrations/gguf/_builder.py]: Fixed in `4e9ab2e`: `_validate_gguf_model` now explicitly accepts `GGUFModel` and returns `GGUFArchitectureAdapter | None`, with type-only imports to avoid runtime cycles.
